darwin/future/core/client.py

```diff
@@ -0,0 +1,203 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Callable, Dict, List, Optional, Union, overload
+
+import requests
+from pydantic import BaseModel, validator
+from requests.adapters import HTTPAdapter, Retry
+
+from darwin.future.core.types.query import Query
+from darwin.future.data_objects.darwin_meta import Team
+from darwin.future.exceptions.client import NotFound, Unauthorized
+
+HTTPMethod = Union[Callable[[str], requests.Response], Callable[[str, dict], requests.Response]]
+
+
+class Config(BaseModel):
+    """Configuration object for the client
+
+    Attributes
+    ----------
+    api_key: Optional[str], api key to authenticate
+    base_url: pydantic.HttpUrl, base url of the API
+    default_team: Optional[Team], default team to make requests to
+    """
+
+    api_key: Optional[str]
+    base_url: str
+    default_team: Optional[Team]
+
+    @validator("base_url")
+    def validate_base_url(cls, v: str) -> str:
+        v = v.strip()
+        if not v.endswith("/"):
```

Comment:
I see what you mean about Pydantic validators here, they are a bit wordy at times, and the mutation of the value in a validator feels almost kind of dirty.

No change suggested, just WRT prev convo.

Reply:
You can seperate them into seperate functions and they become a one line call, I use that elsewhere. Mutation within the validator is the pydantic way, they combine both validation and conversion, if the function throws it's invalid and if it doesn't you mutate into the format you want.